src/kpro_connection.erl

```diff
@@ -0,0 +1,598 @@
+%%%
+%%%   Copyright (c) 2014-2018, Klarna AB
+%%%
+%%%   Licensed under the Apache License, Version 2.0 (the "License");
+%%%   you may not use this file except in compliance with the License.
+%%%   You may obtain a copy of the License at
+%%%
+%%%       http://www.apache.org/licenses/LICENSE-2.0
+%%%
+%%%   Unless required by applicable law or agreed to in writing, software
+%%%   distributed under the License is distributed on an "AS IS" BASIS,
+%%%   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%%%   See the License for the specific language governing permissions and
+%%%   limitations under the License.
+%%%
+
+-module(kpro_connection).
+
+%% API
+-export([ all_cfg_keys/0
+        , get_api_vsns/1
+        , get_tcp_sock/1
+        , init/4
+        , loop/2
+        , request_sync/3
+        , request_async/2
+        , start/3
+        , stop/1
+        , debug/2
+        ]).
+
+%% system calls support for worker process
+-export([ system_continue/3
+        , system_terminate/4
+        , system_code_change/4
+        , format_status/2
+        ]).
+
+-export_type([ config/0
+             , connection/0
+             ]).
+
+-include("kpro.hrl").
+
+-define(DEFAULT_CONNECT_TIMEOUT, timer:seconds(5)).
+-define(DEFAULT_REQUEST_TIMEOUT, timer:minutes(4)).
+-define(SIZE_HEAD_BYTES, 4).
+
+%% try not to use 0 corr ID for the first few requests
+%% as they are usually used by upper level callers
+-define(SASL_AUTH_CORRID, ((1 bsl 31) - 1)).
+-define(API_VERSIONS_CORRID, ((1 bsl 31) - 2)).
```

Comment:
I'm confused.  In an earlier file you state that correlation IDs are 24-bit integers, but here you clearly use 31-bit integers.  What's the truth here?

Reply:
I believe you got the impression from a comment that is deleted in this change:
in `include/kpro_private.hrl` : 
```
-%% correlation IDs are 32 bit signed integers.	
-%% we use 24 bits only, and use the highest 5 bits to be redudant with API key	
-%% and next 3 bits with API version	
-%% so that the decoder may decode the responses without the need of an extra	
-%% correlation ID to API key association.
```

Since in this new version, kafka added so many new APIs and new versions.
in the new implementation, API names and versions are kept in sent requests (State#requests) together with the caller pid and call reference.